core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java

```diff
@@ -1071,53 +1463,125 @@ private void mergeAndLogEntry(FileEntry entry, int mergeCounter) {
          throw new PersistenceException("Could not add new merged entry", e);
       }
    }
+
    @Override
-   public void purge(Executor threadPool, final PurgeListener task) {
-      long now = timeService.wallClockTime();
+   public Publisher<MarshallableEntry<K, V>> purgeExpired() {
+      UnicastProcessor<MarshallableEntry<K, V>> processor = UnicastProcessor.create();
+      blockingManager.runBlocking(() -> blockingPurgeExpired(processor), "sfs-purgeExpired");
+      return processor;
+   }
+
+   private void blockingPurgeExpired(UnicastProcessor<MarshallableEntry<K, V>> processor) {
+      try {
+         long now = timeService.wallClockTime();
+         for (int segment = 0; segment < actualNumSegments; segment++) {
+            List<KeyValuePair<Object, FileEntry>> entriesToPurge;
+            long stamp = resizeLock.readLock();
+            try {
+               Map<K, FileEntry> segmentEntries = getSegmentEntries(segment);
+               if (segmentEntries == null)
+                  continue;
+
+               entriesToPurge = collectExpiredEntries(now, segmentEntries);
+            } finally {
+               resizeLock.unlockRead(stamp);
+            }
+
+            purgeExpiredEntries(now, processor, entriesToPurge);
+         }
+
+         // Disk space optimizations
+         processFreeEntries();
+      } catch (Throwable t) {
+         processor.onError(t);
+      } finally {
+         processor.onComplete();
+      }
+   }
+
+   private void purgeExpiredEntries(long now, UnicastProcessor<MarshallableEntry<K, V>> processor,
+                                    List<KeyValuePair<Object, FileEntry>> entriesToPurge) {
+      entriesToPurge.sort(Comparator.comparingLong(kvp -> kvp.getValue().offset));
+      for (ListIterator<KeyValuePair<Object, FileEntry>> it = entriesToPurge.listIterator(); it.hasNext(); ) {
+         KeyValuePair<Object, FileEntry> next = it.next();
+         FileEntry fe = next.getValue();
+         if (fe.isExpired(now)) {
+            it.set(null);
+
+            // Safe to unlock because the entry was locked collectExpiredEntries
+            // FIXME We cannot load the metadata here or the lifespan check in CallInterceptor will fail
```

Comment:
This sounds like a bug in the CallInterceptor that you linked. Do you think you can log a JIRA for that as well?

Reply:
I created ISPN-13295 and updated the comment.

I was hoping you'd hit the same problem and fix it in https://github.com/infinispan/infinispan/pull/9524 so I wouldn't have to do anything ;)

Reply:
I did not, I am guessing the SFS tests just covers something extra :)